Make ClassfileParser more efficient, lazy, and less cycle prone. [ci: last-only]

retronym opened the pull request:
- Don't intern non-names as Names (e.g. string constants, descriptors,
    and generic descriptors
  - Use type completers for method signatures of Java methods, bringing
    them into line with the way we enter Scala-originated method symbols.
    This should be beneficial for performance, and also avoids cycles
    that have been reported by users.
  - But, be sure that these lazy types don't hold a reference to the
    entire classfile. Instead, make our constant pool implementation
    use only the slice of the classfile containing the pool.

Fixes scala/scala-dev#287

retronym committed:
Avoid Names for descriptors, generic sigs, and string constants

We can just keep these are short-lived Strings, rather than interning
them into the Name table for the entire lifetime of Global.

retronym committed:
Eagerly read from the constant pool as a basis for lazy types java class/method

I've used lazy types for field/method/class infos, which is analagous to what we
do in `Unpickler` for scala originated types.  We read all data needed by the
inner class table and the type completers from the pool eagerly, but still be
lazy about interning strings to Names and completion of the field/method
types themselves.

This fixes some long standing spurious cyclic errors:

Manually tested with:

```
$ scalac -cp $(coursier fetch -q -p com.datastax.cassandra:dse-driver:1.0.0) test.scala
test.scala:2: error: illegal cyclic reference involving class Cluster
	new com.datastax.driver.dse.DseCluster.Builder()
                                               ^
one error found

$ /code/scala/build/quick/bin/scalac -cp $(coursier fetch -q -p com.datastax.cassandra:dse-driver:1.0.0) test.scala

$ cat test.scala
class Test {
	new com.datastax.driver.dse.DseCluster.Builder()
}
```

```
class Test {
  new com.datastax.driver.dse.DseCluster.Builder()
}
```

It subsumes `-Ybreak-cycles`. I was to compile `pos/cycle{-jsoup,}` without the experimental
`-Ybreak-cycles`. I've removed the entire `-Ybreak-cycles` option and supporting code.

It also subsumes the existing implementation related to cycle avoidance
(`queueLoad` / `raiseLoaderLevel`.)

Descriptor, signature, constant, and param names are made available to the field/method lazy
completer which can use them if and when the symbol is completed.

After classfile parsing the constant pool is discarded (some entries are of course captured by
the lazy type completers.)

Fixes scala/bug#3809

retronym commented:
/cc @smarter this is the branch we were discussing yesterday.

retronym commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -187,6 +189,7 @@ abstract class ClassfileParser {
         }
       }
     }
+    private val in1: AbstractFileReader = new AbstractFileReader(java.util.Arrays.copyOfRange(in.buf, initBp, in.bp))
```
Alternatively, we could expose `AbstractFileReader.getChar` and friends as static methods that accept the `byte[]` as an extra parameter. That would avoid the need to have a full `AbstractFileReader` here which is clunky.

lrytz replied:
Maybe we can have a subtype of `AbstractFileReader` that takes an offset parameter, to avoid passing it everywhere explicitly `in.x(y - initBp)`

lrytz replied:
And maybe mutate the array in the `in`  reader instead of having two readers around and setting one to null?

retronym replied:
These details aren't relevant after my later refactorings.

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -519,7 +519,8 @@ abstract class ClassfileParser {
 
       // attributes now depend on having infos set already
-      parseAttributes(clazz, classInfo)
+      val paramNames = ListBuffer[Name](null)
```
unused val

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1313,6 +1261,76 @@ abstract class ClassfileParser {
   }
+  final class JavaSignatureTypeCompleter(name: Name, jflags: JavaAccFlags, index: Int) extends LazyType {
+    override def complete(sym: symbolTable.Symbol): Unit = {
+      var info = pool.getType(sym, index)
```
can be `val`

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1313,6 +1261,76 @@ abstract class ClassfileParser {
+
+  final class JavaGenericSignatureTypeCompleter(sig: String, isVarargs: Boolean) extends LazyType {
+    override def complete(sym: symbolTable.Symbol): Unit = {
+      var info = sigToType(sym, sig)
```
can be `val`

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1313,6 +1261,76 @@ abstract class ClassfileParser {
+      case Some(ParamNames(paramNames)) =>
+        if (sym.hasRawInfo && sym.isMethod) {
+          val params = sym.rawInfo.params
+          val paramNames1 = paramNames.takeRight(params.length)
```
add a comment why. also the [spec](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.24) does not reeeally mandate the number of parameter names to match the number of parameters, so things would be shifted here.

Since the `complete` method (which drops the outer) runs right before `addParamNames`, we can just pass the fact as a parameter to `addParamNames`, no?

retronym replied:
Agreed, I'll rework this. We should probably grab both the descriptor and the signature (if it exists) and the param names (if they exist) in `parseAttributes`, and pass all of that into the completer. The completer can then figure out which param names to use.

The param names are specced to correspond to the descriptor signature (perhaps only a prefix of it, though).

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -868,20 +837,16 @@ abstract class ClassfileParser {
         case tpnme.ConstantValueATTR =>
           val c = pool.getConstant(u2)
-          val c1 = convertTo(c, symtype)
+          val c1 = convertTo(c, sym.initialize.info.resultType)
```
`.initialize` is redundant, no?

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -868,20 +837,16 @@ abstract class ClassfileParser {
+          val c1 = convertTo(c, sym.initialize.info.resultType)
           if (c1 ne null) sym.setInfo(ConstantType(c1))
-          else devWarning(s"failure to convert $c to $symtype")
+          else devWarning(s"failure to convert $c to ${sym.initialize.info.resultType}")
```
also `.initialize`

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1313,6 +1261,76 @@ abstract class ClassfileParser {
+  final class JavaGenericSignatureTypeCompleter(sig: String, isVarargs: Boolean) extends LazyType {
+    override def complete(sym: symbolTable.Symbol): Unit = {
+      var info = sigToType(sym, sig)
+      sym.setInfo(if (isVarargs) arrayToRepeated(info) else info)
```
It seems this wasn't there before for the `sigToType` case, is this a bugfix?

retronym replied:
This used to be called in `parseMethod` after type was initially created based on the descriptor and after it was recreated based on the signature (if there was one). So it did apply in both cases.

retronym committed:
Print name table size after compilation under -verbose

retronym commented:
@lrytz I've implemented my idea to gather the descriptor, signature, param names, and constant indices and lazily/holistically use these in the completer.

I've also added logging for the name table size under `-verbose`, we can see the effect of the changes herein:

```
7415b728da Print name table size after compilation under -verbose
Name table size after compilation: 102264 chars
70551b21f2 Avoid Names for descriptors, generic sigs, and string constants
Name table size after compilation: 81370 chars
7133ed1dad (HEAD) Read parameter names from constant pool lazily
Name table size after compilation: 54732 chars
```

retronym commented:
/synch

retronym commented:
I've briefly tried this change out within scala/compiler-benchmark.

`hot -psource=vector -pextraArgs=-Ystop-after:typer -prof gc -f1` shows that `gc.alloc.rate.norm` reduce 37.7MiB to 36.36MiB (0.96x).

Execution time for the same (measured over more forks) also slightly improves:

```scala
[info] HotScalacBenchmark.compile                          e3df10a  -Ystop-after:typer       false  2.13.0-pre-7415b72-SNAPSHOT    vector  sample  2674   67.530 ± 0.182  ms/op
[info] HotScalacBenchmark.compile                          e3df10a  -Ystop-after:typer       false  2.13.0-pre-616d7ba-SNAPSHOT    vector  sample  2765  65.334 ± 0.149  ms/op
```

I'd expect this improvement to be mostly experienced by small compilation batches like `vector`; in larger runs the classloading time isn't that significant.

What I haven't done is measured the retained size of the compiler before and after this change. While we're saving on some name/type/symbols, we are keeping the `byte[]` of the constant pool of java classes around.

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1261,27 +1252,41 @@ abstract class ClassfileParser {
-      if (name == nme.CONSTRUCTOR)
-        info match {
+      if (constantIndex != -1) {
+        val c = pool.getConstant(constantIndex)
```
Unused

lrytz reviewed (commented):
Looks good; there are still some (older) review comments to address.

retronym commented:
@lrytz I've tried the refactoring to trim the stored `byte[]` internally within `AbstractFileReader`, using a transparent offset. I pulled the thread a bit further and extracted an interface `DataReader`. A second implementation of this uses a `ByteBuffer` internally, as well as a single instance of `DataInputStream` for UTF decoding.

This opens the door to:
  - using memory mapped byte buffers to read classfiles from disk (although I'm not sure its a good idea).
  - having a small pool of reusable `ByteBuffer`-s for classfile reading (like we already do for source file reading), knowing that we can release the buffer back to the pool after we've `trim`-ed the constant pool.

TODO:
  - [ ] benchmark the changes
  - [ ] if `isScala || isScalaRaw`, we don't actually need the `ConstantPool` anymore, so we should avoid the `trim` method.

mkeskells commented on src/reflect/scala/reflect/io/PlainFile.scala:
```diff
@@ -42,6 +46,16 @@ class PlainFile(val givenPath: Path) extends AbstractFile {
   override def output = givenPath.toFile.outputStream()
   override def sizeOption = Some(givenPath.length.toInt)
+  override def toByteBuffer: ByteBuffer = {
+    val chan = java.nio.file.Files.newByteChannel(file.toPath, util.EnumSet.of(StandardOpenOption.READ))
```
probably worth stuffing the set in a constant

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ByteBufferDataReader.scala:
```diff
@@ -0,0 +1,87 @@
+    require(offset == 0, "Cannot trim more than once")
+    data.position(start)
+    data.limit(end)
+    data.compact()
```
It seems this doesn't actually release any memory:

```
scala> val b = java.nio.ByteBuffer.wrap(Array.tabulate[Byte](128)(_.toByte))
b: java.nio.ByteBuffer = java.nio.HeapByteBuffer[pos=0 lim=128 cap=128]

scala> b.position(7)
res85: java.nio.Buffer = java.nio.HeapByteBuffer[pos=7 lim=128 cap=128]

scala> b.limit(16)
res86: java.nio.Buffer = java.nio.HeapByteBuffer[pos=7 lim=16 cap=128]

scala> b.compact()
res87: java.nio.ByteBuffer = java.nio.HeapByteBuffer[pos=9 lim=128 cap=128]

scala> b.position(0)
res88: java.nio.Buffer = java.nio.HeapByteBuffer[pos=0 lim=128 cap=128]

scala> b.array.length
res89: Int = 128
```

retronym replied:
Ha, good catch. I'll just make `data` a var, like in the other implementation.

lrytz commented:
LGTM otherwise, feel free to squash or not.

retronym commented:
`java.util.HashMap` has a constant pool 7918 bytes out of a classfile size of 17591. That's a bigger proportion that I intuited. So its not clear to me what's the best trade off, maybe it would be better to eagerly parse it all into basic data types (like String), so we don't need to capture the `byte[]` at all for later use in the lazy completers.

retronym commented:
I'm trying out the approach reading everything we need from the constant pool when setting up the completers. I've also tried using a completer for the class info type itself, which nicely replaces the ad-hoc cycle avoidance implemented in `raiseLoaderLevel`.

The parts that I'm not happy with is the way that we still eagerly resolve references to enum constants in annotations, which could trigger extra classfile parsing. I think it would be better to have a pass that only extracts the enum owner name and enum name, and then resolves that to the actual symbol lazily (when the annotated methods is completed, or even later when its annotations are first forced). Same story for class constants. We could eagerly construct a tree representing the applied annotation, and wrap that in `LazyAnnotationInfo(() => treeToAnnotation(tree)))`.

SethTisue commented:
closing for inactivity, can be reopened/revived whenever

smarter commented:
Does it really need to be closed ? That seems like it'll make it much more likely it will be forgotten about. It certainly decreases the visibility of future changes for anyone looking at the PR queue.

SethTisue commented:
we use the queue to track work items that are actively being pursued. therefore we like to keep the number of open PRs manageable. anything over 50 or so leads to widespread forgetting and neglecting, PR-queue-wide, and lately we've been up around 80–120, it's too much.

it's ultimately Adriaan's call whether to adopt a different policy, but that's how we currently manage the queue and have for some time.

ijuma commented:
This has the WIP label, could that be used to filter out WIP PRs?

SethTisue commented:
It can, but it also matters what shows up in the default view, in practice, that’s what everybody’s going to look at 80% of the time.

Ticket for continuing discussion around this: https://github.com/scala/scala-dev/issues/577

retronym commented:
I've revived this by rebasing atop the latest 2.13.x.

retronym committed:
Reuse the buffer for classfile reading

Classfile parsing does re-enter when we're reading package objects
or classfiles for things like `scala/native.class`.

But for the most part the prior refactorings mean that we typically
only parse a single classfile at a time, and as such we can profit
from a one-element cache for the buffer to read this into.

retronym committed:
Avoid using Names for fully qualified class names

There is no good reason for these dotted names to be
Names and stick around in the name table. Let's use
short lived strings instead.

Reduces the name table by 5% in terms of entries and 10%
in terms of characters when compiling src/scalap/**/*.scala

retronym commented:
[Benchmark results](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&var-scalaSha=83a0dca87a%7C4483d732e4%7C04a775c258%7C116ca94cec%7C88a9377eaf%7C3d190c09aa%7Caf5d493117%7Cc0a768285a%7C2acea19f75%7C229257c1fe%7C367b98798d&var-source=scala&var-source=scalap&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40&from=1550462471545&to=1551238575771) for the latest commit vs 2.13.x.

retronym committed:
Resort to old-school git HEAD detection in the build

retronym commented:
I've cleaned this up. Its sprouted a few partest failures which I believe are noise and will trim back tomorrow. Then it's ready for re-review.

The allocation rate improvement now looks a bit more impressive than [last time I measured](https://github.com/scala/scala/pull/6653#issuecomment-393052393). That's due to caching the buffer from the classfile bytes.

```
$ (cd compiler-benchmark; for v in 2.13.0-pre-4483d73-SNAPSHOT 2.13.0-pre-f0d5231-SNAPSHOT; do echo $v; sbt 'set scalaVersion in compilation := "'$v'"' 'hot -psource=vector -pextraArgs=-Ystop-after:typer -prof gc -f1' | grep HotScalacBenchmark.compile.*gc.alloc.rate.norm; done)
/code/compiler-benchmark
2.13.0-pre-4483d73-SNAPSHOT
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                  ../corpus          e3df10a  -Ystop-after:typer       false  2.13.0-pre-4483d73-SNAPSHOT    vector  sample    10  37997324.033 ±    2810.295    B/op
2.13.0-pre-f0d5231-SNAPSHOT
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                  ../corpus          e3df10a  -Ystop-after:typer       false  2.13.0-pre-f0d5231-SNAPSHOT    vector  sample    10  33613725.336 ±    2787.168    B/op
```

plokhotnyuk commented on src/compiler/scala/tools/nsc/symtab/classfile/ReusableDataReader.scala:
```diff
@@ -0,0 +1,149 @@
+  private[this] var size = 0
+  private[this] val reader: DataInputStream = {
+    val stream = new InputStream {
+      override def read(): Int = try {
```
It would be better to override also the default implementation of `public int read(byte b[], int off, int len)` because it is used in `readFully` immediately and in `readUTF` through `readFully`, so it will save a lot of CPU cycles.

retronym replied:
Good idea!

retronym committed:
Override the bulk version of read for efficiency

lrytz commented on src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala:
```diff
@@ -294,11 +294,11 @@ abstract class SymbolLoaders {
     }
   }
-
+  private val classFileDataReader: ReusableInstance[ReusableDataReader] = new ReusableInstance[ReusableDataReader](() => new ReusableDataReader())
```
If once a large classfile is loaded, the reader's data array will stay at this size for the lifetime of the `Global` instance (right?). Do you think this is fine?

retronym replied:
Yes, that's correct.

In the SBT use case the `Global` lifetime is limited to a single compilation batch, so I think that's okay.

For the presentation compiler, the "high water mark"-style leak could be felt more, especially if multiplied by a large number of `Global` instances for sub-projects. 

We could revisit this cache to:
  - purge after each `Run`
  - make the cache pan-`Global` (would require making `ReusableInstance` race-proof)
  - disable with a policy setting

lrytz reviewed (approved):
Only one question. Very nice work!

adriaanm commented:
@retronym needs a rebase -- then ready to merge?

retronym commented:
```
[error] java.io.IOException: PUT operation to URL https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots;build.timestamp=1552915259957/org/scala-lang/scalap/2.13.0-pre-4686e30-SNAPSHOT/scalap-2.13.0-pre-4686e30-SNAPSHOT.pom failed with status code 503: Service Unavailable
[error] 	at org.apache.ivy.util.url.AbstractURLHandler.validatePutStatusCode(AbstractURLHandler.java:82)
```

retronym commented:
/rebuild

retronym commented:
```
[0m[[0m[0minfo[0m] [0m[0mCompiling 562 Scala sources and 168 Java sources to /home/travis/build/scala/scala/build/quick/classes/library ...[0m
[0m[[0m[31merror[0m] [0m[0m[34m-- [E007] Type Mismatch Error: /home/travis/build/scala/scala/src/library/scala/collection/mutable/TreeSet.scala:197:42 [0m[0m
[0m[[0m[31merror[0m] [0m[0m[31m197 |[0m        [33mnew[0m [35mTreeSet[E][0m(RB.fromOrderedKeys(it, r.size))[0m
[0m[[0m[31merror[0m] [0m[0m    |                                          [31m^^[0m[0m
[0m[[0m[31merror[0m] [0m[0m    |                                          Found:    Iterator[[1m[31mInt[0m[1m[31m][0m[1m[31m([0m[1m[31mit[0m[1m[31m)[0m[0m
[0m[[0m[31merror[0m] [0m[0m    |                                          Required: Iterator[[1m[32mE[0m][0m
[0m[[0m[33mwarn[0m] [0m[0m[34m-- Warning: /home/travis/build/scala/scala/src/library/scala/concurrent/ExecutionContext.scala:163:2 [0m[0m
[0m[[0m[33mwarn[0m] [0m[0m[31m163 |[0m  [33mfinal[0m [33mobject[0m [35mparasitic[0m [33mextends[0m [35mExecutionContextExecutor[0m [33mwith[0m [35mBatchingExecutor[0m {[0m
[0m[[0m[33mwarn[0m] [0m[0m    |  [31m^^^^^[0m[0m
[0m[[0m[33mwarn[0m] [0m[0m    |  [33mfinal[0m modifier is redundant for objects[0m
[0m[[0m[33mwarn[0m] [0m[0m[34m-- Warning: /home/travis/build/scala/scala/src/library/scala/concurrent/ExecutionContext.scala:169:2 [0m[0m
[0m[[0m[33mwarn[0m] [0m[0m[31m169 |[0m  [33mfinal[0m [33mobject[0m [35mImplicits[0m {[0m
[0m[[0m[33mwarn[0m] [0m[0m    |  [31m^^^^^[0m[0m
[0m[[0m[33mwarn[0m] [0m[0m    |  [33mfinal[0m modifier is redundant for objects[0m
[0m[[0m[33mwarn[0m] [0m[0mtwo warnings found[0m
[0m[[0m[31merror[0m] [0m[0mone error found[0m
[0m[[0m[31merror[0m] [0m[0m(library / Compile / [31mcompileIncremental[0m) Compilation failed[0m
[0m[[0m[31merror[0m] [0m[0mTotal time: 41 s, completed Mar 18, 2019 2:09:46 PM[0m

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
travis_time:end:284a4d10:start=1552918096453802714,finish=1552918186548062956,duration=90094260242
[0K[31;1mThe command "sbt -Dscala.build.compileWithDotty=true library/compile" exited with 1.[0m
```

retronym commented:
@smarter I'd recommend porting these changes over to dotty.
  - [ ] Avoid use of `Name` table entries for signatures, descriptors, fully-qualiified classnames
  - [ ] Use lazy class and member types for Java-defined symbols from the classpath, making sure that the completers don't capture references to the `byte[]`, but instead to names/strings (constant pool entries)
  - [ ] reused the `byte[]` for parsing, allowing for reentrancy (which we still get the internals of classfile parser need to force the info of some other package, which needs to first parse its package object class.

smarter commented:
@retronym Agreed, this PR has been on my mental todo-list for a while :). Just opened an issue to keep track of it: https://github.com/lampepfl/dotty/issues/6136
